Port to 1.21.5

AnonymousHacker1279 opened the pull request:
This PR ports to MC 1.21.5 on NeoForge 21.5.65-beta. It also includes a related MDG update to 2.0.88. 

Since this is a fairly major porting update, I decided to also take time to clean up most of the workspace, and to optimize where possible. As such, several API methods have changed signatures, but I believe this is acceptable considering the change in MC version. Hopefully it should be easier to maintain in the future. 

## Rundown
- Removed post TP sound event mixin; according to #9, it has been inactive and broken anyways since 1.20. After my own investigation it seems complicated to re-implement, and I did not see a good mixin entry point.
- Moved the commented test code to a separate test sourceset that is always active during development, and included a second test portal.
- Allowed `CustomPortalBuilder()` to be created directly instead of needing a `beginPortal()` method
- Removed custom event wrappers for pre/post teleport events; these are handled with simple consumers/functions.
- Removed `SHOULDTP` enum and replaced with a simple boolean for the pre teleport event.
- Allow frame testers to be directly accessed and set, removing the need to register them and reference via `ResourceLocation`s.

## Removed Files:
- `WorldRendererMixin.java` (related to the post TP sound event)
- `PerWorldPortals.java` (unused)
- `CustomPortalParticle.java` (unused, and can be trivially implemented by mods if necessary)
- `CPAEvent.java` (unnecessary; replaced with `Function`s)
- `CPASoundEventData.java` (unused)
- `ColorUtil.java` (`getColorForBlock` method unused, `getColorFromRGB` used in one place so it was inlined)
- `SHOULDTP.java` (unnecessary; replaced with a boolean)
- `CustomPortalApiRegistry.java` (mostly moved into the base mod file)
- `CustomPortalsModClient.java` (block color handlers event moved into `ClientEventSubscriber.java`, client setup event for render types removed as they are now set directly in the model JSON)
- `customportalparticle.json` (unused)

## Added Files:
- `DimensionLink.java` (represents two `DimensionalBlockPos` objects, used for saving to disk)
- `TestMod.java` (under new test sourceset, basically what got moved out of `CustomPortalsMod.java`)
- `ClientEventSubscriber.java` (client-specific mod events)
- `NeoEventSubscriber.java` (game mod events)
- `package-info.java` (to each package with nullability annotations)

## Other Notes

I did keep the recently added `CustomPortalRegistrationEvent`, however, I am a little confused on why it exists. Since it only calls `build()` on a `CustomPortalBuilder` instance, it effectively acts as a proxy for the `FMLCommonSetupEvent` it is posted from. I feel it makes more sense to just create and build these objects during the setup event directly, and it's what I've been doing in my own mod.

AnonymousHacker1279 committed:
Fire post teleport event

Fire the post teleport event again, as it didn't seem to be hooked up to anything previously.

Update dependencies: NeoForge to 21.1.61, MDG to 2.0.31-beta, Parchment to 2024.07.78, and Gradle to 8.9.

AnonymousHacker1279 committed:
Merge branch '1.21.1' of https://github.com/AnonymousHacker1279/customportalapi-reforged into 1.21.1

AnonymousHacker1279 committed:
Fix ResourceLocationExceptions

Fix ResourceLocationException errors when loading portal locations due to incorrect parsing.

AnonymousHacker1279 committed:
Merge branch '1.21.1' of https://github.com/AnonymousHacker1279/customportalapi-reforged into 1.21.1

AnonymousHacker1279 committed:
Port to 1.21.5

Port to 1.21.5. Includes related dependency updates.

The post teleport sound event mixin has been removed; according to #9, it does not seem to have been working anyways since 1.20.

AnonymousHacker1279 committed:
Include test mod

Move the commented test code from the main source set to a test set.

Slight restructuring of save data classes.

AnonymousHacker1279 committed:
Fix for non-fluid portal ignition sources

Fix logic for the isWater/isLava methods throwing with non-fluid portal ignition sources.

Add a second test portal.

Removed registration methods for in/post ambience sounds as neither are hooked up anymore.

AnonymousHacker1279 committed:
Begin cleaning up API surface

Clean up API surface and simplify where possible.

Deleted custom portal particle and per world portals classes as both were unused.

AnonymousHacker1279 committed:
Cleanup pre/post teleport events

Remove custom event wrappers for pre/post teleport events.

Remove unused CPASoundEventData class.

Remove unused methods in PortalIgnitionSource; the TODO for withCondition has been present since the initial push in 2023.

AnonymousHacker1279 committed:
Remove need to register frame testers

Allow frame testers to be directly accessed, removing the need to register them and reference via ResourceLocation.

AnonymousHacker1279 committed:
More cleanup

Various cleanups across the rest of the codebase.

Added package-info.java files to each package.

AzureDoom commented:
Hey thanks for the port! I'll review this here likely this upcoming week and move it into it's own branch as well. Thanks again for this!

AnonymousHacker1279 committed:
Implement sound functions

Add the ability to set travel/trigger/ambient sounds to portals.

AnonymousHacker1279 committed:
Allow custom portal particles to be set

Allow custom portal particles to be set without having to subclass CustomPortalBlock.

AnonymousHacker1279 commented:
Because I am incapable of sleeping (it is now past 4:30AM here), I decided to also work on the sound stuff.

Travel sounds have been implemented with a custom packet based on what I use in my own mods, trigger sounds have been implemented with a small mixin to `LocalPlayer#handlePortalTransitionEffect`, and ambient sounds just by adding a hook into the custom portal block.

Additionally, I added a builder method for specifying custom particle types on the portal block.

AnonymousHacker1279 committed:
Fix post teleport event

Fix the post teleport event not firing after implementing the PostTeleportTransition sound changes.

AnonymousHacker1279 committed:
Fix crashes on dedicated server

Fix crashes on dedicated server due to client classes being loaded.

AzureDoom commented:
Everything is looking good here!
